sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -30,211 +30,382 @@ import org.apache.spark.util.Utils
+      case _ => false
     }
   }
 
-  private def skipForShortcut(expr: Expression): Expression = {
-    if (skipForShortcutEnable) {
-      // The subexpression may not need to eval even if it appears more than once.
-      // e.g., `if(or(a, and(b, b)))`, the expression `b` would be skipped if `a` is true.
-      expr match {
-        case and: And => and.left
-        case or: Or => or.left
-        case other => other
-      }
-    } else {
-      expr
+  private def updateWithExpr(
+      expr: Expression,
+      evalCount: Int,
+      condEvalCount: Double): Boolean = {
+    require(evalCount >= 0 && condEvalCount >= 0d)
+
+    inflated = false
+    val wrapper = ExpressionEquals(expr)
+    val map = getMapByHeight(wrapper.height)
+    map.get(wrapper) match {
+      case Some(es) =>
+        es.evalCount += evalCount
+        es.condEvalCount += condEvalCount
+        true
+      case _ =>
+        map(wrapper) = ExpressionStats(expr)(evalCount, condEvalCount, 0, 0d)
+        false
     }
   }
 
-  // There are some special expressions that we should not recurse into all of its children.
-  //   1. CodegenFallback: it's children will not be used to generate code (call eval() instead)
-  //   2. ConditionalExpression: use its children that will always be evaluated.
-  private def childrenToRecurse(expr: Expression): Seq[Expression] = expr match {
-    case _: CodegenFallback => Nil
-    case c: ConditionalExpression => c.alwaysEvaluatedInputs.map(skipForShortcut)
-    case other => skipForShortcut(other).children
+  private def getMapByHeight(height: Int) = {
+    val index = height - 1
+    while (maps.size <= index) {
+      maps += mutable.Map.empty
+    }
+    maps(index)
   }
 
-  // For some special expressions we cannot just recurse into all of its children, but we can
-  // recursively add the common expressions shared between all of its children.
-  private def commonChildrenToRecurse(expr: Expression): Seq[Seq[Expression]] = expr match {
-    case _: CodegenFallback => Nil
-    case c: ConditionalExpression => c.branchGroups
-    case _ => Nil
+  // Iterate expressions from parents to children and fill inflated `realEvalCount`s and
+  // `realCondEvalCount`s from explicitly added `evalCount`s and `condEvalCount`s.
+  private def inflate() = {
+    if (!inflated) {
+      maps.reverse.foreach { map =>
+        map.foreach {
+          case (_, es) => inflateExprState(es)
+          case _ =>
+        }
+      }
+      inflated = true
+    }
   }
 
-  private def supportedExpression(e: Expression) = {
-    !e.exists {
-      // `LambdaVariable` is usually used as a loop variable, which can't be evaluated ahead of the
-      // loop. So we can't evaluate sub-expressions containing `LambdaVariable` at the beginning.
-      case _: LambdaVariable => true
+  private def inflateExprState(exprStats: ExpressionStats): Unit = {
+    val expr = exprStats.expr
+    if (!expr.isInstanceOf[LeafExpression] || allowLeafExpressions) {
```

Comment:
what will go wrong if we always allow leaf expressions?

Reply:
We don't need to store leaf expressions when we use `EquivalentExpressions` for CSE as leafs don't make sense to evaluate in advance. `addExprTree()` didn't recurse to leafs in the previous version but now that `addExprTree()` doesn't recurse, we need this flag.